### PR TITLE
BSA: add native SM120 blk128 forward and FA4-style optimization

### PR DESCRIPTION
**Pull request by tiffany940107**
## Before submitting

- [x] I agree to license this contribution under the terms of [LICENSE.txt](https://github.com/NVIDIA/cudnn-frontend/blob/develop/LICENSE.txt).
- [x] I ran `pre-commit run` and committed any formatting changes.
- [x] I reviewed the Hard Rules in the Python and test directory `AGENTS.md` files.
- [x] Labels: `cat-feature`, `area:sparse_attention`, and `orig-external`.

## Affected area

FE OSS CuTeDSL block-sparse attention forward kernels, Python dispatch,
numerical tests, benchmarks, and documentation.

## Summary

Add dedicated **native KV128** forward kernels for SM120 FP16/BF16,
with QK/V head dimensions 128. Both kernels consume blk128 sparse metadata
directly. There is **no KV128-to-KV64 metadata expansion and no blk64-kernel
call in the native128 production path**.

- General native blk128 kernel for variable counts, explicit valid block
  sizes, partial Q/KV tiles, and MHA/GQA.
- Fixed-top-k FA4-style specialization: register-resident Q, dedicated K/V
  TMA load warp, independent K/V synchronization, early sparse-address
  preparation, and shared-memory reuse for the output epilogue.
- Four-fold producer/compute sparse-loop unrolling and warpgroup-uniform
  register donation.
- Balance an underfilled final scheduling wave by giving only eligible final
  Q128 work to two Q64 CTAs. **Each CTA still loads complete KV128 blocks and
  uses its parent Q block's original metadata.** This is Q-work scheduling,
  not the KV-block lowering in #1010. There is one kernel launch and no new
  scratch buffer; SM-count planning runs during JIT tracing, not cache hits.
- A current-checkout benchmark against PyTorch's cuDNN dense SDPA backend,
  with performance gates, and expanded FP16/BF16 numerical coverage.
  Historical comparison scripts and raw measurement reports are archived
  locally, not included in the current PR file diff. Usage commands live
  in the forward test module docstring; no separate benchmark README is
  included.

"FA4-style" describes a scheduling/storage strategy adapted to SM120; this
does not call an external FlashAttention-4 package. BF16 Q/K/V remain BF16,
with FP32 accumulation and BF16 probabilities. No FP8/INT8 quantization,
polynomial exponential, or thresholded normalization is enabled.

## Why / relationship to #1010

This is an independent native-kernel follow-up to the closed logical-blk128
adapter PR #1010. It does not reopen that PR or replace its branch history.
The intended workload is BF16 BHSD `[1, 8, 142720, 128]`, sparse block size
128, at 15% or 20% density. Related to #442.

### Historical same-device comparison with native blk64 and #1010

One RTX PRO 6000 Blackwell **Server Edition**, SM120, 188 SMs;
PyTorch `2.13.0+cu130`. All three paths shared Q/K/V and exactly the same
selected KV tokens. Native64 mask preparation was excluded from timing;
the archived #1010 API included its per-call metadata conversion.

Each independent run used 12 warmups and 102 CUDA-event samples per path per
case, cycling equally through all six launch-order permutations. Compilation
and validation were excluded. The second run measured:

| Actual density | Pattern | Native blk64 | PR #1010 | Native blk128 | vs blk64 | vs #1010 |
| ---: | --- | ---: | ---: | ---: | ---: | ---: |
| 14.9776% | strided | 35.4504 ms | 35.4771 ms | 31.6873 ms | 1.1188x | 1.1196x |
| 14.9776% | local | 35.6599 ms | 35.6702 ms | 31.7743 ms | 1.1223x | 1.1226x |
| 20.0000% | strided | 47.6678 ms | 47.6842 ms | 42.4655 ms | 1.1225x | 1.1229x |
| 20.0000% | local | 49.3349 ms | 49.3656 ms | 43.7443 ms | 1.1278x | 1.1285x |

Across both runs, the gain is approximately **12%-13% throughput**, or
**10.6%-11.4% lower latency**, against those paths. These results use the
`e0602027` GPU kernel checkpoint; subsequent publication work adds portable
benchmarking, documentation, and a host-side DSL-version guard. The review
follow-up adds kernel docstrings but leaves executable kernel statements and
numerical tolerances unchanged; locally archived historical timing artifacts
retain their measured source hashes. Performance tables stay in this PR
description; the historical comparison harnesses are archived locally.
The retained test-file commands measure and validate the current checkout,
not historical revisions.

The separate additional-speedup goal uses an **already optimized FA4-style
native128** baseline, `9869b9b6`. Current measured ratios against that baseline
are only 1.0079x-1.0114x. **The additional 1.05x goal has not been achieved.**
This is historical local A/B evidence; the saved-kernel comparison harness is
no longer included in the submission.
Do not combine these different baselines or mix earlier Workstation Edition
absolute timings with the Server Edition table.

## API and compatibility impact

- No public signature change; SM120 still defaults to blk64. Opt in with
  `sparse_block_size=128`.
- Forward-only, QK/V dimensions 128, FP16/BF16, BHSD/BSHD and unpacked
  per-query-head metadata. SM120 still requires `kv_splits=1`,
  `use_clc=None`, and `pack_gqa=None` or `False`.
- Fixed top-k, `q2k_block_nums=None`, `block_sizes=None`, and full physical
  KV blocks use the FA4-style specialization. Other supported blk128 cases
  use the general native kernel.
- The FA4-style path requires public CuTe DSL >= 4.7.0. Older versions receive
  a version-specific error **before** importing its experimental DSL APIs.
  The package-wide downstream dependency floor is unchanged.
- Multi-GPU communication/compute-overlap prototypes remain in the separate
  Block-Sparse-Attention repository and are outside this change.

## Testing

Run the retained BSA suite directly; no legacy source archive or historical
revision is required:

```bash
cd test/python
CUDA_VISIBLE_DEVICES=0 python -m pytest -q fe_api/bsa
```

Final local result after removing the historical comparison harnesses:
**52 passed, 19 skipped**. The 14 removed three-path cases exercised only that
deleted benchmark, explaining the reduction from 66 at `77c0d2e2`; production
kernel correctness tests are unchanged. This covers the SM120 CuTe BSA
path, not other GPU architectures or backend Graph API CI; the local frontend
backend-version probe was unavailable. Unsupported configurations skip.

Both historical comparison benchmarks and their harness-only tests have
been removed from the submission and archived locally. The benchmark directory
now contains only `benchmark_sm120_blk128.py`, which measures this checkout
against PyTorch's cuDNN dense SDPA backend. Its implementation is unchanged
and its `--help` smoke check passed. Usage is in the forward test module
docstring linked below.

```bash
pre-commit run
pre-commit run --from-ref "$(git merge-base upstream/develop HEAD)" --to-ref HEAD
git diff --check upstream/develop...HEAD
```

Applicable Black, Black Jupyter, and SPDX checks passed; no changed C++ files
required clang-format.

Accuracy and regression evidence:

- FP16/BF16 top-k 1-5, partial Q/KV, GQA, query-dependent noncontiguous
  selection, and all-tail/full/mixed/unsplit scheduling cases.
- SM-count cache-hit regression was observed RED, then GREEN after its fix.
- The old-DSL import-guard regression was observed RED before the new guard
  and GREEN afterward; tests for the specialization skip on an old DSL.
- Kernel helper docstrings were added without changing executable statements.
  The latest cleanup changes only the forward test module's usage docstring
  in retained Python files; an AST comparison confirms its executable tests
  are unchanged. Kernel sources and the current-version benchmark are
  byte-for-byte unchanged by this cleanup.
- Historical full-workload comparisons found PR #1010 and native64 full O/LSE
  tensors bitwise identical. Native128 versus native64 was numerical, not bitwise:
  maximum full O/LSE differences were `2.4414e-4` / `3.8147e-6`; full-tensor
  comparisons and sampled FP32-reference checks passed.
- Separately, native128 versus its `9869b9b6` baseline was bitwise identical
  for full O/LSE tensors at both densities and address patterns.
- No production accuracy tolerance was relaxed. At Q=12,161 and top-k=1,
  an exploratory BF16 FP32-reference check with `atol=1e-3, rtol=1e-2`
  exceeded its bound at two of 3,113,216 output elements. The unchanged
  control reproduced the same errors; full candidate/control O/LSE tensors
  were bitwise equal. This is not counted as a passing strict-reference test.
- Rejected experimental kernels and screening notes are archived locally.
  In particular, a normalization-removal diagnostic is not attention and
  is not counted toward the additional 1.05x target.

## Review and reproduction entry points

- [Single-GPU correctness tests and dense benchmark usage](https://github.com/tiffany940107/cudnn-frontend/blob/bsa-sm120-native-blk128-fa4-style/test/python/fe_api/bsa/test_BSA_attention_forward.py)
- [BSA API and support matrix](https://github.com/tiffany940107/cudnn-frontend/blob/bsa-sm120-native-blk128-fa4-style/docs/fe-oss-apis/bsa.md)

Performance tables are retained in this PR description rather than separate
tracked reports. Raw JSON reports, the four removed benchmark Markdown files,
and both historical comparison harnesses are archived locally. The retained
benchmark prints fresh current-checkout sparse/dense timings. Usage is
documented directly in the forward test file. The existing BSA API documentation
is retained, and `AGENTS.md` is unchanged from the upstream base.
Machine IPs, hostnames, GPU UUIDs, private absolute paths, raw profiler captures,
and rejected prototype sources are not published. The archived #1010 adapter
was only a historical local benchmark baseline, never a production fallback
for native128.

## Summary by CodeRabbit

- **New Features**
  - Added SM120 support for 128×128 block-sparse attention, including optimized forward execution and optional FA4-style processing.
  - Added configurable sparse block sizes with validation across supported GPU architectures.

- **Benchmarks**
  - Added SM120 benchmarks for block-sparse attention and comparisons across supported execution paths, including accuracy, performance, density, and speedup reporting.

- **Documentation**
  - Expanded setup, usage, requirements, and performance guidance for SM120 testing.
  - Removed an outdated optimization summary link.

- **Tests**
  - Added coverage for SM120 correctness, layouts, metadata variants, edge cases, and benchmark safety checks.

### CODE REVIEW

**Comment by tiffany940107** [comment]
The required cat-feature, area:sparse_attention, and orig-external labels are applied. The merge-requirements check currently reports a missing Milestone and Project assignment; those fields are left for maintainers as requested by AGENTS.md. Local pre-commit checks over the complete PR diff passed, and the final SM120 BSA regression result is 66 passed / 19 skipped. No GPU-architecture-wide CI result is claimed.

File: benchmark/bsa/benchmark_sm120_three_paths.py [hunk 1]
```diff
@@ -0,0 +1,264 @@
...
+        "warmup": args.warmup,
+        "repeats": args.repeats,
+        "source_sha256": hashes,
+        "candidate_revision": subprocess.check_output(["git", "-C", str(CHECKOUT), "rev-parse", "HEAD"], text=True).strip(),
+        "pr1010_revision": PR1010_REVISION,
+        "method": "Same-process six-permutation CUDA-event public-wrapper timing; native64 preprocessing excluded; PR1010 conversion included; compile and validation excluded.",
+        "records": records,
+    }
+    args.json.parent.mkdir(parents=True, exist_ok=True)
+    args.json.write_text(json.dumps(report, indent=2) + "\n")
```
[the remark this answers is not in the document]
Comment (tiffany940107) [reply benchmark/bsa/benchmark_sm120_three_paths.py]: Fixed in 3c6e02dd. The report now uses `Path.open("x", encoding="utf-8")`, so a file created after argument validation is not truncated.

Added `test_three_path_cli_smoke[True]`, which creates a competing report after CLI validation, runs the real benchmark path, expects `FileExistsError`, and verifies both preservation of the competing file and cleanup of the isolated legacy imports. The regression was observed RED with the original `write_text()` and GREEN after this fix. All 13 three-path harness cases pass; the complete BSA run with the pinned legacy archive enabled reports 67 passed and 19 skipped.

File: benchmark/bsa/SM120_THREE_PATHS.md [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+Across the two runs, current native blk128 reaches 1.1188x-1.1278x versus native blk64 and 1.1196x-1.1285x versus PR #1010. This is approximately 12%-13% more throughput or 10.6%-11.4% lower latency. Absolute timings drift; each ratio compares paths interleaved in the same process on the same device.
+
+The measured PR #1010 versus native64 median difference is about 0.010-0.033 ms, while current native128 saves about 5.1-5.6 ms at 20% density. Most of this improvement is therefore associated with native128 kernel execution, not elimination of metadata conversion. The small difference also includes wrapper/launch differences and measurement variation; it is not a separately isolated conversion-kernel timer.
+
+The earlier 0.8%-1.1% gain used the already optimized FA4-style native128 `9869b9b6` as its baseline. It is not the baseline used here, and these results do not claim that the additional 1.05x goal against `9869b9b6` has been achieved.
+
+## Validation
+
+- After promotion to the portable public harness, the complete BSA suite passed
+  **65 tests with 19 skips**, with all 12 new harness cases enabled. These
```
[the remark this answers is not in the document]
Comment (tiffany940107) [reply benchmark/bsa/SM120_THREE_PATHS.md]: Corrected in 3c6e02dd in both `SM120_THREE_PATHS.md` and `SM120_REVIEW_SUMMARY.md`.

The records now distinguish the actual checkpoints: 65 passed / 19 skipped at `9211954a` (portable harness), then 66 passed / 19 skipped at `daa7e0a8` (added DSL-version guard). The current reviewer-fix run is **67 passed / 19 skipped**, because the concurrent-output CLI regression adds one case. All 13 three-path harness cases are enabled. Historical timing samples and source hashes are preserved, not rewritten.

**Comment by tiffany940107** [comment]
Review follow-up pushed:
- `3c6e02dd`: exclusive JSON report creation, a RED-then-GREEN concurrent-output CLI regression, and explicitly versioned validation counts.
- `ac24eb06`: docstrings for touched functions/classes, including native blk128 kernel helpers and benchmark/test contracts.

For the docstring coverage warning, local AST inspection finds docstrings on all 72 touched functions/classes. CodeRabbit will recompute its own coverage. Removing docstrings produces identical ASTs for all eight Python files in the documentation-only commit; kernel arithmetic, scheduling, and numerical tolerances are unchanged. Historical benchmark samples and measured source hashes are preserved.

Final local validation on SM120:
- Three-path harness: **13 passed**.
- Targeted SM120 forward tests after docstring additions: **25 passed**.
- Complete BSA directory with the pinned legacy archive enabled: **67 passed, 19 skipped**.
- Staged-file and full-PR-diff pre-commit hooks: **PASS**.
- Diff whitespace and review-change privacy checks: **PASS**.

Both inline findings have replies with implementation and regression details. The PR description is updated to 67 passed / 19 skipped. These results cover local SM120 CuTe BSA, not other architectures or backend Graph API CI. Milestone/Project fields were not modified by this follow-up; the latest merge-requirements check now passes.

**Comment by tiffany940107** [comment]
PR scope cleanup in `12cf84da`:
- Removed the two historical JSON reports from the current PR file diff after byte-for-byte local archival. No timing values or summary tables were changed.
- Reverted only this PR's added `AGENTS.md` line; the existing upstream file is preserved and no longer appears in the PR diff.
- Removed stale JSON links and updated the benchmark docs and PR description. Portable benchmark scripts, report generation, regression tests, and performance tables remain.
- No kernel, benchmark-script, or test-source changes.

Validation after removing the reports: **13 three-path tests passed**; full-PR-diff pre-commit, documentation links, unchanged-table checks, and local archive integrity checks passed. The prior full BSA checkpoint remains **67 passed / 19 skipped** at `ac24eb06`; this cleanup did not rerun the full BSA suite.

This is a normal follow-up commit, not a history rewrite.

**Comment by tiffany940107** [comment]
Documentation consolidation pushed in `964fafb6`, including removal of the benchmark README as requested.

- Removed all four standalone benchmark Markdown files from the current PR diff after byte-for-byte local archival.
- Usage, requirements, correctness commands, dense comparison, pinned PR #1010 archive preparation, and saved-kernel A/B commands now live in the module docstrings of the three relevant test files.
- Benchmark `--help` and the missing-revision error point to those test files. The existing BSA API/support documentation is retained without stale report links.
- Performance tables and necessary accuracy/performance limitations remain in this PR description, including the unachieved additional 1.05x goal and the pre-existing strict-reference exceedance. No new performance result is claimed.

Validation: **68 passed, 19 skipped** for the complete BSA directory with the pinned archive enabled; **16 passed** for the two benchmark test files independently. The added missing-revision hint regression was observed RED then GREEN. All three benchmark help commands, shell syntax of documented commands, archive integrity, stale-link/privacy checks, and staged/full-PR-diff pre-commit passed.

Production kernel files and benchmark timing/accuracy logic are unchanged. This is a normal follow-up commit; archived documentation remains recoverable.

**Comment by tiffany940107** [comment]
Removed the historical FA4 paired benchmark and its companion test file in `77c0d2e2`, as requested:
- `benchmark/bsa/benchmark_sm120_blk128_pair.py`
- `test/python/fe_api/bsa/test_sm120_blk128_pair_benchmark.py`

The files were archived locally and verified byte-for-byte before removal. The remaining test usage and PR links no longer reference them. The current-version dense benchmark, three-path comparison, production kernels, and numerical test bodies are unchanged. The four standalone benchmark Markdown files remain removed.

Validation: **66 passed, 19 skipped** across the complete BSA directory with the pinned legacy archive enabled. The count decreases by exactly the two deleted paired-harness tests; no production kernel correctness case was removed. Staged and full-PR-diff pre-commit, whitespace, archive-integrity, remaining-reference, and privacy checks passed. The current-version benchmark's `--help` also passed.

The PR description now distinguishes the historical A/B evidence from the benchmark code retained in the submission.

**Comment by tiffany940107** [comment]
Removed the remaining historical three-path benchmark and its test harness in e64b25ff:
- `benchmark/bsa/benchmark_sm120_three_paths.py`
- `test/python/fe_api/bsa/test_sm120_three_paths_benchmark.py`

Also removed their stale usage/legacy-archive references. The historical performance table in the PR description is unchanged; both deleted files are archived locally. `benchmark/bsa/` now contains only `benchmark_sm120_blk128.py` for measuring the submitted checkout against the dense backend.

Kernel sources, the retained benchmark, and executable correctness tests are unchanged. The full local BSA suite passed with **52 passed, 19 skipped**, without a legacy archive; the reduction from 66 is solely the 14 removed harness-only cases. The remaining benchmark's `--help` smoke check, staged-file pre-commit, full-PR-diff pre-commit, and `git diff --check` all passed. No raw results or machine-specific connection details are included.

**Review by jiayus-nvidia — APPROVED** [review]

**Comment by jiayus-nvidia** [comment]
@cudnn-ci-bot run oss
